### PR TITLE
sort when encoding coordinates for deterministic outputs

### LINKED ISSUE #8026: non-deterministic ordering within "coordinates" attribute written by .to_netcdf
### What is your issue?

Under the assumption that deterministic output is preferred whenever feasible, I'd like to point out that the variable names written into "coordinates" attributes with `.to_netcdf` are not ordered deterministically. For pipelines that depend on file hashes to validate dependencies, this can be a real headache.

Consider the dataset `xarray.Dataset({"x": ((), 0)}, coords={"a": 0, "b": 0})`. The NetCDF file XArray writes will include either:
```
variables:
	int64 x ;
		x:coordinates = "a b" ;
	int64 a ;
	int64 b ;
```
or
```
variables:
	int64 x ;
		x:coordinates = "b a" ;
	int64 a ;
	int64 b ;
```

My review of [`_encode_coordinates`](https://github.com/pydata/xarray/blob/52f5cf1fc521891008c4c968cee0b389211481c0/xarray/conventions.py#L631) leads me to think the behavior results from collecting names in a `set`. I'd be happy to offer a PR to make the coordinates attribute deterministic. I am not aware of a CF convention regarding any ordering, but would research and follow if it exists. If not, then I would probably `sort` at [L701](https://github.com/pydata/xarray/blob/52f5cf1fc521891008c4c968cee0b389211481c0/xarray/conventions.py#L701) and [L722](https://github.com/pydata/xarray/blob/52f5cf1fc521891008c4c968cee0b389211481c0/xarray/conventions.py#L722).

### PR DESCRIPTION
**Pull request by itcarroll**
- [x] Closes #8026
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- ~~[ ] New functions/methods are listed in `api.rst`~~

The PR changes `conventions._encode_coordinates` to add `sorted` during the creation of "coordinates" strings *de novo*. It does not touch user-specified coordinates attributes or encodings.

The PR adds a test, but also changes two pre-existing tests that allowed for non-deterministic ordering of the coordinates string.

In reviewing the netCDF data model and CF convenstions, I confirmed that there is no requirement on the ordering of names in the coordinates attribute. I also learned that the global coordinates attribute created by XArray for non-dimension coordinates that are not associated with a variable is not compliant with CF conventions, as was [discussed in 2014](https://mailman.cgd.ucar.edu/pipermail/cf-metadata/2014/007571.html). In 2021, CF-Conventions 1.9 added ["Domain Variables"](http://cfconventions.org/cf-conventions/cf-conventions.html#domain-variables), which appear to provide a CF compliant way to handle this situation. I will likely open an enhancement issue to weigh making that change.

### CODE REVIEW

File: xarray/tests/test_conventions.py [hunk 1]
```diff
@@ -129,9 +129,9 @@ def test_multidimensional_coordinates(self) -> None:
         foo1_coords = enc["foo1"].attrs.get("coordinates", "")
         foo2_coords = enc["foo2"].attrs.get("coordinates", "")
         foo3_coords = enc["foo3"].attrs.get("coordinates", "")
-        assert set(foo1_coords.split()) == {"lat1", "lon1"}
-        assert set(foo2_coords.split()) == {"lat2", "lon2"}
-        assert set(foo3_coords.split()) == {"lat3", "lon3"}
+        assert foo1_coords == "lon1 lat1"
```
Comment (dcherian) [inline xarray/tests/test_conventions.py:132]: Why is this not `"lat1 lon1"` if it is sorted? (and same below e.g. `"time lon lat"`)

Comment (itcarroll) [reply xarray/tests/test_conventions.py:132]: User-specified coordinates attributes (or encodings further below) are not sorted, only internally assigned coordinates get sorted. User-specified attributes are already deterministic.

**Review by dcherian — APPROVED** [review]
Thanks @itcarroll

File: doc/whats-new.rst [hunk 2]
```diff
@@ -96,6 +96,9 @@ Internal Changes
 
 - :py:func:`as_variable` now consistently includes the variable name in any exceptions
   raised. (:pull:`7995`). By `Peter Hill <https://github.com/ZedThree>`_
+- :py:func:`encode_dataset_coordinates` now sorts coordinates automatically assigned to
+  `coordinates` attributes during serialization (:pull:`8034` by
```
Comment (headtr1ck) [inline doc/whats-new.rst:100]:
```suggestion
  `coordinates` attributes during serialization (:issue:`8026`, :pull:`8034`).
```

File: doc/whats-new.rst [hunk 3]
```diff
@@ -96,6 +96,9 @@ Internal Changes
...
+  `Ian Carroll <https://github.com/itcarroll>`_).
```
Comment (headtr1ck) [inline doc/whats-new.rst:101]:
```suggestion
  `By Ian Carroll <https://github.com/itcarroll>`_.
```

**Review by headtr1ck — APPROVED** [review]
Run into this issue as well in the past, thanks for fixing this (and reporting it in the first place)!

**Pull request merged**